src/python/py/models/builder.py

```diff
@@ -21,16 +22,19 @@
 
 
 class Model:
-    def __init__(self, config, io_dtype, onnx_dtype, ep, cache_dir, extra_options):
-        self.context_length = config.max_position_embeddings
-        self.original_context_length = config.original_max_position_embeddings if hasattr(config, "original_max_position_embeddings") else config.rope_scaling["original_max_position_embeddings"] if hasattr(config, "rope_scaling") and hasattr(config.rope_scaling, "original_max_position_embeddings") else config.max_position_embeddings
+    def __init__(self, config, io_dtype, onnx_dtype, ep, cache_dir, extra_options): 
+        self.context_length = config.seq_length if hasattr(config, "seq_length") else config.max_position_embeddings
+        self.original_context_length = config.original_max_position_embeddings if hasattr(config, "original_max_position_embeddings") else config.rope_scaling["original_max_position_embeddings"] if hasattr(config, "rope_scaling") and hasattr(config.rope_scaling, "original_max_position_embeddings") else self.context_length
         self.window_size = config.sliding_window if hasattr(config, "sliding_window") else -1  # default is -1 in GroupQueryAttention kernel
-        self.intermediate_size = config.intermediate_size
+        self.intermediate_size = config.intermediate_size if hasattr(config, "intermediate_size") else config.ffn_hidden_size
         self.hidden_size = config.hidden_size
         self.num_kv_heads = config.num_key_value_heads if hasattr(config, "num_key_value_heads") else config.num_attention_heads
+        self.num_kv_heads = config.multi_query_group_num if hasattr(config, "multi_query_group_num") else self.num_kv_heads
+        self.kv_channels = config.kv_channels if hasattr(config, "kv_channels") else self.num_kv_heads
+        self.multi_query_attention = config.multi_query_attention if hasattr(config, "multi_query_attention") else False
```

Comment:
Multi-query attention can be identified with `self.num_kv_heads == 1`. Since it is only used once within the `ChatGLMModel` class and it is a small check, I don't think this is needed.

Reply:
Since, the ChatGLMModel using GQA with self.num_kv_heads = 2 currently [Here](https://huggingface.co/THUDM/chatglm3-6b/blob/91a0561caa089280e94bf26a9fc3530482f0fe60/config.json#L28), I think we can use the following check to choose GQA or MHA. Please let me know if we should try something else?
```python
self.attention_attrs["op_type"] = "GroupQueryAttention" if self.num_attn_heads > self.num_kv_heads else self.attention_attrs["op_type"]

Reply:
There is a later section to handle the selection of the attention op type.

https://github.com/microsoft/onnxruntime-genai/blob/f76867300d35a9e3ee66a2779b965700ea256e73/src/python/py/models/builder.py#L229-L247

This section already handles choosing between GQA and MHA as the attention op. The set of valid GQA configurations covers most precision + execution provider combinations. Since GQA is a superset of MHA, we can use GQA when `self.num_attn_heads != self.num_kv_heads` or `self.num_attn_heads == self.num_kv_heads`. Thus, I don't think we need to worry about selecting the attention op for ChatGLM.